src/main/extension/extension_install.cpp

```diff
@@ -347,46 +349,82 @@ static unique_ptr<ExtensionInstallInfo> InstallFromHttpUrl(DBConfig &config, con
+
 		if (install_info && !install_info->etag.empty()) {
 			headers.insert({"If-None-Match", StringUtil::Format("%s", install_info->etag)});
 		}
-	}
-
-	auto res = cli.Get(url_local_part.c_str(), headers);
 
-	if (install_info && res && res->status == 304) {
-		return install_info;
-	}
+		res = cli.Get(url_local_part.c_str(), headers);
+		if (install_info && res && res->status == 304) {
+			return install_info;
+		}
 
-	if (!res || res->status != 200) {
-		// create suggestions
-		string message;
-		auto exact_match = ExtensionHelper::CreateSuggestions(extension_name, message);
-		if (exact_match && !ExtensionHelper::IsRelease(DuckDB::LibraryVersion())) {
-			message += "\nAre you using a development build? In this case, extensions might not (yet) be uploaded.";
+		if (res && res->status == 200) {
+			// success!
+			break;
 		}
+		// failure - check if we should retry
+		bool should_retry = false;
 		if (res.error() == duckdb_httplib::Error::Success) {
-			throw HTTPException(res.value(), "Failed to download extension \"%s\" at URL \"%s%s\" (HTTP %n)\n%s",
-			                    extension_name, url_base, url_local_part, res->status, message);
+			switch (res->status) {
+			case 408: // Request Timeout
+			case 418: // Server is pretending to be a teapot
+			case 429: // Rate limiter hit
+			case 500: // Server has error
+			case 503: // Server has error
+			case 504: // Server has error
+				should_retry = true;
+				break;
+			default:
+				break;
+			}
 		} else {
-			throw IOException("Failed to download extension \"%s\" at URL \"%s%s\"\n%s (ERROR %s)", extension_name,
-			                  url_base, url_local_part, message, to_string(res.error()));
+			// always retry on duckdb_httplib::Error::Error
+			should_retry = true;
+		}
+		retry_count++;
+		if (!should_retry || retry_count >= MAX_RETRY_COUNT) {
+			// if we should not retry or exceeded the number of retries - bubble up the error
+			string message;
+			auto exact_match = ExtensionHelper::CreateSuggestions(extension_name, message);
+			if (exact_match && !ExtensionHelper::IsRelease(DuckDB::LibraryVersion())) {
+				message += "\nAre you using a development build? In this case, extensions might not (yet) be uploaded.";
+			}
+			if (res.error() == duckdb_httplib::Error::Success) {
+				throw HTTPException(res.value(), "Failed to download extension \"%s\" at URL \"%s%s\" (HTTP %n)\n%s",
+				                    extension_name, url_base, url_local_part, res->status, message);
+			} else {
+				throw IOException("Failed to download extension \"%s\" at URL \"%s%s\"\n%s (ERROR %s)", extension_name,
+				                  url_base, url_local_part, message, to_string(res.error()));
+			}
 		}
+		// retry
+		// sleep first
+		uint64_t sleep_amount =
+		    static_cast<uint64_t>(static_cast<double>(RETRY_WAIT_MS) * pow(RETRY_BACKOFF, retry_count - 1));
+		std::this_thread::sleep_for(std::chrono::milliseconds(sleep_amount));
```

Comment:
```suggestion
#ifndef DUCKDB_NO_THREADS
		std::this_thread::sleep_for(std::chrono::milliseconds(sleep_amount));
#endif
```
Using of sleep_for might not be there always (wasm, cough cough). I think we will end up with a duckdb_threads header to wrap relevant logic in a way that works cross platforms, here I will just add the #ifdef otherwise compilation will fail.

Actually, probably even just having a:
#include "duckdb_threads.h"


that inside is like:
```c++
#ifndef DUCKDB_NO_THREADS
#include <threads>
#endif
```
can already solve some problems, and move towards abstracting some primitives.
But that's not for this PR.

Reply:
Is any of this code even present/used in the WASM client? I thought this was all removed?

Reply:
Code is there, compiled but not used.
You can also leave it in, you decide.

Reply:
I added in the ifdef